<center>

$\Huge \textbf{Universidad Nacional Autónoma de México}$  
$\Huge \textbf{Facultad de Ciencias}$  
<p align="center">
  <img src="https://www.icat.unam.mx/wp-content/uploads/2021/11/Copia-de-LogoUNAM.-Azul.-Fondo-transparente.png" alt="UNAM" width="200"/>
</p>

<hr style="height:3px; background-color:#0B6E4F; border:none;"/>


$\LARGE \textbf{Inteligencia Artificial}$  

$\Large \textit{Laboratorio 1.7}$  


\begin{array}{rl}
\textbf{Docente:} & Dra. Jessica Sarahi Méndez Rincón \\[6pt]
\textbf{Alumno:} & emiliano, Kaleb Jimenez Rivera \\[6pt]
\textbf{Fecha de realización:} & 14/09/2026
\end{array}

</center>

In [ ]:
%load_ext rpy2.ipython

In [ ]:
%%R
mi_vector <- c(1, 2, 3, 4, 5)
mean(mi_vector)

[1] 3


# Código en Python (Medición de tiempo)
En esta celda definimos la prueba en Python utilizando time y numpy.

In [ ]:
import numpy as np
import time

def evaluar_tiempo_python():
    print("--- Ejecutando en Python ---")
    start_time = time.time()

    # Generar matrices aleatorias de 2000x2000
    A = np.random.rand(2000, 2000)
    B = np.random.rand(2000, 2000)

    # Multiplicación de matrices
    C = np.dot(A, B)

    end_time = time.time()
    tiempo_total = end_time - start_time
    print(f"Tiempo de ejecución en Python: {tiempo_total:.4f} segundos\n")
    return tiempo_total

# Ejecutar la función
tiempo_py = evaluar_tiempo_python()

--- Ejecutando en Python ---
Tiempo de ejecución en Python: 0.6826 segundos



# Código en R (Medición de tiempo)

Usamos la etiqueta %%R al inicio de la celda. Nota cómo usamos <- para guardar el tiempo en R, pero al final podemos regresar esa variable a Python usando la bandera -o tiempo_r.

In [ ]:
%%R -o tiempo_r
evaluar_tiempo_r <- function() {
  cat("--- Ejecutando en R ---\n")
  start_time <- Sys.time()

  # Generar matrices aleatorias de 2000x2000
  A <- matrix(runif(2000 * 2000), nrow = 2000, ncol = 2000)
  B <- matrix(runif(2000 * 2000), nrow = 2000, ncol = 2000)

  # Multiplicación de matrices
  C <- A %*% B

  end_time <- Sys.time()
  tiempo_total <- as.numeric(difftime(end_time, start_time, units = "secs"))
  cat(sprintf("Tiempo de ejecución en R: %.4f segundos\n\n", tiempo_total))
  return(tiempo_total)
}

# Ejecutar la función y asignarla a la variable que saldrá a Python
tiempo_r <- evaluar_tiempo_r()

--- Ejecutando en R ---
Tiempo de ejecución en R: 0.9442 segundos



# Comparación Final
Como exportamos tiempo_r en la celda anterior, ahora podemos comparar ambos resultados directamente en una celda de Python para ver cuál fue más veloz:

In [ ]:
print("=== COMPARATIVA DE TIEMPOS ===")
print(f"Python: {tiempo_py:.4f} segundos")
print(f"R:      {tiempo_r[0]:.4f} segundos") # rpy2 devuelve una lista/vector, tomamos el primer elemento

if tiempo_py < tiempo_r[0]:
    ganador = "Python"
    diferencia = tiempo_r[0] / tiempo_py
else:
    ganador = "R"
    diferencia = tiempo_py / tiempo_r[0]

print(f"\n🏆 {ganador} fue el más rápido en esta prueba por un factor de {diferencia:.2f}x.")

=== COMPARATIVA DE TIEMPOS ===
Python: 0.6826 segundos
R:      0.9442 segundos

🏆 Python fue el más rápido en esta prueba por un factor de 1.38x.


# Un detalle técnico a tener en cuenta
Al ejecutar esto en Colab, notarás que ambos lenguajes son extremadamente rápidos para esta tarea (suele tomar menos de un segundo). Esto se debe a que por debajo, tanto numpy en Python como las funciones nativas de R en Colab utilizan librerías de álgebra lineal altamente optimizadas (como BLAS/LAPACK) que aprovechan al máximo los hilos del procesador de Google.

#Conclusiones finales del estudiante


En esta práctica vimos como tanto Python como R, son posibles de utilizar dentro de un mismo notebook, además de que comparten la misma filosofía sobre medir tiempo de cómputo. Creo que lo más útil que se aprendió en esta práctica fue como mediante una bandera -o tiempo_r se permitió regresar un resultado calculado en R al entorno de Python y mediante esta técnica se pudo comparar ambos tiempos en una sola celda. También podemos observar que R no devuelve un número suelto, a comparación de Python, si no un vector, es por esto que tuvimos que acceder al tiempo mediante índices. En general, pudimos contrastar ambos lenguajes al multiplicar dos matrices de 2000 × 2000. En el caso de Python lo hicimos con np.random.rand y no.dot, midiendo con la biblioteca de time; y por su parte, con él, rehusamos reunir y un operador especial %*% y el método de tiempo del sistema.


Ahora bien, basándonos en la prueba que hicimos, se obtuvo 0.682 segundos en Python contra 0.944 segundos en R por lo que Python fue más rápido. Sin embargo, conviene tomar este resultado con pinzas por dos razones. La primera es que sólo ejecutamos una simple operación, y esto no generaliza el comportamiento de un lenguaje. La segunda razón es que el tiempo medido incluye también la generación de números aleatorios, no sólo la multiplicación de las matrices. De la misma forma, ambos tiempos quedaron por debajo de un segundo, y tenemos que notar que Python y sus operaciones matemáticas están hechas sobre el lenguaje C, lo cual, en la mayoría de operaciones se queda en tiempos de ejecución muy cortos a comparación de R, el cual usa librerías como PLUCH y LPACK. No podemos decir que este Test nos da el mejor lenguaje para ejecutar operaciones matemáticas, pues ambos usan biblio, bibliotecas de terceros. Entonces lo que estaríamos midiendo sería qué tanta aprovechan estas bibliotecas y las optimizaciones para utilizarlas.Podemos concluir que ambos son lenguajes muy importantes para proyectos relacionadas con IA y ciencia de datos, y que ninguno de estos dos está peleados si no se complementan para la construcción de mejor software.

<hr/>
<footer style="text-align:center; font-size:12px; color:gray;">
© 2026 UNAM Facultado de Ciencias – Todos los derechos reservados

</footer>